File: vanilla/vanilla-task-repetition.js
class Task {
    constructor(description) {
      this.description = description;
    }
  
    async run() {
      return this.description;
    }
  }
  
  class CustomProcess {
    constructor(tasks = []) {
      this.tasks = tasks;
      this.executionHistory = [];
    }
  
    async run() {
      const results = [];
      for (const task of this.tasks) {
        const result = await task.run();
        results.push(result);
        this.executionHistory.push(task.description);
        console.log(`Task executed: ${task.description}`);
      }
      return results;
    }
  
    add_task(task, repetitions = 1) {
      for (let i = 0; i < repetitions; i++) {
        this.tasks.push(task);
      }
    }
  }
  
  class Agent {
    constructor() {}
  
    async executeProcess(process) {
      return process.run();
    }
  }
  async function main() {
    const task1 = new Task("Hello");
    const task2 = new Task("world");
    const myProcess = new CustomProcess([task2]);
    myProcess.add_task(task1, 2);  // Add task1 twice
    myProcess.add_task(task2);
    const agent = new Agent();
    const results = await agent.executeProcess(myProcess);
    console.log("Execution history:", myProcess.executionHistory.join(" "));
  }
  main();

File: vanilla/vanilla-task-repetition.py
class Task:
    def __init__(self, description):
        self.description = description

    async def run(self):
        return self.description

class CustomProcess:
    def __init__(self, tasks):
        self.tasks = tasks
        self.execution_history = []

    async def run(self):
        for task in self.tasks:
            result = await task.run()
            self.execution_history.append(task.description)
            print(f"Task executed: {task.description}")
        return self.execution_history

    def add_task(self, task, repetitions=1):
        """Adds a task to the process, repeating it if specified."""
        self.tasks.extend([task for _ in range(repetitions)])

class Agent:
    def __init__(self):
        pass

    async def execute_process(self, process):
        return await process.run()

async def main():
    task1 = Task("Task 1")
    task2 = Task("Task 2")

    my_process = CustomProcess([task2])
    my_process.add_task(task1, 2)  # Add task1 twice
    my_process.add_task(task2)

    agent = Agent()

    results = await agent.execute_process(my_process)
    print("Execution history:", results)

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
